### ISSUE TITLE
Working residuals for glmmTMB

### ISSUE THREAD

**Issue by lionel68** [glmmTMB/glmmTMB#776]
Dear glmmTMB developer,

Based on [this definition](https://www.datascienceblog.net/post/machine-learning/interpreting_generalized_linear_models/) it seems that it would be easy to implement working residuals in residuals.glmmTMB:

```r
residuals.glmmTMB <- function(...)
[...]
  res <- switch(type, response = r, pearson = {
    if (is.null(v <- family(object)$variance)) stop("variance function undefined for family ", 
                                                    sQuote(family(object)$family), "; cannot compute", 
                                                    " Pearson residuals")
    vv <- switch(length(formals(v)), v(fitted(object)), v(fitted(object), 
                                                          sigma(object)), stop("variance function should take 1 or 2 arguments"))
    r <- r/sqrt(vv)
    if (!is.null(wts)) {
      r <- r * sqrt(wts)
    }
    r
  }, working = {r / fitted(object)})

```

Is this correct? If needed would be happy to open a pull request for this.

**Comment by bbolker** [comment]
I hadn't thought about this before: I usually think of working residuals as being part of the machinery of iteratively reweighted least squares (IRLS), which `glmmTMB` doesn't use. If you look inside `glm.fit`, the residuals (which are stored as `$residuals` in the returned object, and later extracted & returned by `residuals.glm` are defined as `(y - mu)/mu.eta(eta)`, which differs from the definition given in your link. (I think this can be explained by the fact that `mu.eta(eta)` = `d(mu)/d(eta)` = the derivative of the inverse link function, is the same as `mu(eta)` (the inverse link function itself) for a log-link model like the Poisson given in the example. An example illustrating this with a binomial response (where `mu(eta) != mu.eta(eta)`):

```
library(glmmTMB)
data(cbpp, package = "lme4")
m1_tmb <- glmmTMB(formula = incidence/size ~ period,
                  family = binomial,
                  weights = size,
                  data = cbpp)
m1_glm <- glm(formula = incidence/size ~ period,
              family = binomial,
              weights = size,
              data = cbpp)
r1 <- residuals(m1_glm, type = "working")
p1 <- predict(m1_tmb, type = "link", fast = TRUE)
mm <- family(m1_tmb)$mu.eta
all.equal(residuals(m1_tmb)/mm(p1), r1, tolerance = 1e-6)  ## TRUE
all.equal(residuals(m1_tmb)/fitted(m1_tmb), r1, tolerance = 1e-6)  ## relative difference: 0.1
```
A pull request along these lines would be welcome.

**Comment by lionel68** [comment]
Right, I didn't went further that the link provided and assumed that the definition there works for all glm models, thx for correcting me and implementing working residuals in glmmTMB over the WE!

**Comment by bbolker** [comment]
closed/resolved by 3c7915de93e7
